pkg/providers/equinix.go

```diff
@@ -0,0 +1,142 @@
+package providers
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/google/uuid"
+	"github.com/packethost/packngo"
+)
+
+const (
+	hostPrefix string = "ofcir"
+)
+
+var client *packngo.Client
+var eqOnce sync.Once
+
+type equinixProviderConfig struct {
+	ProjectID string   `json:"projectid"` //project id in equinix
+	Token     string   `json:"token"`     //token for authentication
+	Facility  []string `json:"facility"`  //datacenters to use
+	Plan      string   `json:"plan"`      //server size
+	OS        string   `json:"os"`        //OS to install
+}
+
+type equinixProvider struct {
+	config equinixProviderConfig
+	client *packngo.Client
+}
+
+func EquinixProviderFactory(providerInfo string, secretData map[string][]byte) (Provider, error) {
+	config := equinixProviderConfig{
+		ProjectID: "",
+		Token:     "",
+		Facility:  []string{"da11", "dc10", "dc13", "ny5", "sv15"},
+		Plan:      "c3.small.x86",
+		OS:        "rocky_8",
+	}
+
+	if configJson, ok := secretData["config"]; ok {
+		if err := json.Unmarshal(configJson, &config); err != nil {
+			return nil, err
+		}
+	}
+
+	eqOnce.Do(func() {
+		client = packngo.NewClientWithAuth("packngo lib", config.Token, nil)
+	})
+
+	return &equinixProvider{
+		config: config,
+		client: client,
+	}, nil
+}
+
+func (p *equinixProvider) Acquire(poolSize int) (Resource, error) {
+	resource := Resource{}
+
+	//check how many resources exist and compare to pool size spec
+	//to prevent a bug creating infinite resources
+	deviceList, _, err := p.client.Devices.List(p.config.ProjectID, nil)
+	if err != nil {
+		return resource, fmt.Errorf("Error getting equinix devices: %s", err)
+	}
+	count := 0
+	for _, d := range deviceList {
+		if strings.HasPrefix(d.Hostname, hostPrefix) {
+			count++
+		}
+	}
+	if count >= poolSize {
```

Comment:
we may have a problem here if there are multiple equinix pools all sharing the same "hostPrefix" but pool specific poolSize. I'm ok with fixing this later as it likely wont be a problem right away,

Reply:
Not sure why we would have multiple pools in the same equinix project, but I understand the problem.

Reply:
One reasonable case would be also to have different pools for different equinix configurations, for example the family gen.